### PR TITLE
Gpu support

### LINKED ISSUE #67: Move KData to device
Implement standard pytorch functionality to move data + ktraj + header between cpu and gpu (e.g.  .to(device)). Add flags to select of ktraj and header are moved.

### PR DESCRIPTION
**Pull request by ckolbPTB**
I tested a simple reconstruction pipeline on a GPU. Only needed a few fixes for it to run. 

I have implemented a very basic .cuda() and .cpu() functionality for KTrajectory and KData. If we are more or less happy with this approach I would extend this to IData, CsmData...

I also added tests which are marked "cuda". In order to run only the cuda-tests `pytest -m cuda` can be used.

I added the pytest flag `-m "not cuda"` to pytest.yml such that these tests are excluded from CI.

### CODE REVIEW

**Comment by fzimmermann89** [comment]
Some remarks:

I would try to match pytorch .cuda /.cpu / .to signature. What is the reasoning behind the custom handling of the default device? Can't we just pass on the arguments to the Tensor.cpu / .. function and let pytorch decide ?

We should test that for the operators Operator.cuda() ... works. This is inherited from nn.module. For it to work, all tensors should be registered as buffers or parameters, the commands will be passed tp submodules.

I would maybe also add .clone and .detach to the data classes.

**Comment by ckolbPTB** [comment]
> I would try to match pytorch .cuda /.cpu / .to signature. 

You mean to also add `non_blocking` and `memory_format` to the signature `.cuda(device=None, non_blocking=False, memory_format=torch.preserve_format)`?

> Can't we just pass on the arguments to the Tensor.cpu / .. function and let pytorch decide ?

Yes, true.

**Comment by ckolbPTB** [comment]
Issues/questions to be addressed:

- I get the following error from mypy: 
```
src/mrpro/data/_KData.py:195: error: Unexpected keyword argument "memory_format" for "cuda" of "_TensorBase"  [call-arg]
/home/runner/.cache/pre-commit/repo23gjt902/py_env-python3/lib/python3.10/site-packages/torch/_C/__init__.pyi:2009: note: "cuda" of "_TensorBase" defined here
```

- Not sure of the abstract class `Data` is really necessary

- I did not add a `to()` method to `KData` because I am not sure what it should do as in general `KData.data` and `KData.traj` have different dtypes.

File: src/mrpro/data/_Data.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        device: torch.device | None = None,
+        non_blocking: bool = False,
+        memory_format: torch.memory_format = torch.preserve_format,
+    ) -> Data:
+        """Create copy of object with data in CUDA memory.
+
+        Parameters
+        ----------
+        device
+            The destination GPU device. Defaults to the current CUDA device.
```
Comment (fzimmermann89) [inline src/mrpro/data/_Data.py:51]:
```suggestion
            The destination GPU device. Defaults to the current device.
```
can also be CPU, mps, ...

File: src/mrpro/data/_DcfData.py [hunk 2]
```diff
@@ -91,14 +91,17 @@ def _dcf_2d3d_voronoi(traj: torch.Tensor) -> torch.Tensor:
...
+
         # 2D and 3D trajectories supported
         dim = traj.shape[0]
         if dim not in (2, 3):
             raise ValueError(f'Only 2D or 3D trajectories supported, not {dim}D.')
 
         # Calculate dcf only for unique k-space positions
         traj_dim = traj.shape
-        traj = np.round(traj.numpy(), decimals=UNIQUE_ROUNDING_DECIMALS)
+        traj = np.round(traj.cpu().numpy(), decimals=UNIQUE_ROUNDING_DECIMALS)
```
Comment (fzimmermann89) [inline src/mrpro/data/_DcfData.py]: Maybe we can use the oppertunity to give the input traj  and the numpy traj different names? 
This would allow mypy to consider them as array / tensor not array|tensor

this would also remove `traj_device = traj.device`

**Comment by fzimmermann89** [comment]
reported that tensor.cpu has wrong typing information  at https://github.com/pytorch/pytorch/issues/118386

File: src/mrpro/data/_DcfData.py (hunk 2 above)
Comment (ckolbPTB) [reply src/mrpro/data/_DcfData.py]: I left the input traj as traj because this should be our standard traj tensor.

**Comment by fzimmermann89** [comment]
Maybe we should use a decorator to mark tests that require CUDA? 
this is how its often done in other packages, but TBH I don't know how this is implemented then in pytest.

File: src/mrpro/data/_KData.py [hunk 3]
```diff
@@ -171,3 +171,89 @@ def reshape_acq_data(data):
...
+        elif dtype == torch.complex128:
+            dtype_traj = torch.float64
+        else:
+            raise ValueError(f'dtype {dtype} not supported. Only torch.complex64 and torch.complex128 is supported.')
+
+        return KData(
+            header=self.header,
+            data=self.data.to(
+                device=device, dtype=dtype, non_blocking=non_blocking, copy=copy, memory_format=memory_format
+            ),  # type: ignore [call-overload]
```
Comment (fzimmermann89) [inline src/mrpro/data/_KData.py:215]: is # type: ignore [call-overload] actually  requiered here?

File: src/mrpro/data/_KData.py [hunk 4]
```diff
@@ -171,3 +171,89 @@ def reshape_acq_data(data):
...
+            Otherwise, the argument has no effect.
+        memory_format
+            The desired memory format of returned Tensor.
+        """
+        return KData(
+            header=self.header,
+            data=self.data.cuda(
+                device=device, non_blocking=non_blocking, memory_format=memory_format
+            ),  # type: ignore [call-arg]
+            traj=self.traj.cuda(device=device, non_blocking=non_blocking, memory_format=memory_format),
```
Comment (fzimmermann89) [inline src/mrpro/data/_KData.py:244]: and why is # type: ignore [call-overload] not required here as well?

File: src/mrpro/data/_KData.py [hunk 5]
```diff
@@ -171,3 +171,89 @@ def reshape_acq_data(data):
...
+
+        Parameters
+        ----------
+        memory_format
+            The desired memory format of returned Tensor.
+        """
+        return KData(
+            header=self.header,
+            data=self.data.cpu(memory_format=memory_format),  # type: ignore [call-arg]
+            traj=self.traj.cpu(memory_format=memory_format),
```
Comment (fzimmermann89) [inline src/mrpro/data/_KData.py:258]: why is # type: ignore [call-overload] only required for one of the tensors?
Is mypy not checking the other one ?

Comment (fzimmermann89) [reply src/mrpro/data/_KData.py:258]: ..this would mean that there is any issue with our typing somewhere else.
.data and .traj should behave the same wrt type hints.

File: tests/data/test_idata.py [hunk 6]
```diff
@@ -44,3 +45,26 @@ def test_IData_from_kheader_and_tensor(random_kheader, random_test_data):
...
+    idat_cuda = idat.cuda()
+    assert idat_cuda.data.is_cuda
+
+
+@pytest.mark.cuda
+def test_IData_cpu(random_kheader, random_test_data):
+    """Move IData object to CUDA memory and back to CPU memory."""
+    idat = IData.from_tensor_and_kheader(data=random_test_data, kheader=random_kheader)
+    idat_cpu = idat.cuda().cpu()
+    assert idat_cpu.data.is_cpu
```
Comment (fzimmermann89) [inline tests/data/test_idata.py:70]: is_cpu() ?

File: tests/data/test_idata.py [hunk 7]
```diff
@@ -44,3 +45,26 @@ def test_IData_from_kheader_and_tensor(random_kheader, random_test_data):
...
+    idat_complex128 = idat.to(dtype=torch.complex128)
+    assert idat_complex128.data.dtype == torch.complex128
+
+
+@pytest.mark.cuda
+def test_IData_cuda(random_kheader, random_test_data):
+    """Move IData object to CUDA memory."""
+    idat = IData.from_tensor_and_kheader(data=random_test_data, kheader=random_kheader)
+    idat_cuda = idat.cuda()
+    assert idat_cuda.data.is_cuda
```
Comment (fzimmermann89) [inline tests/data/test_idata.py:62]: can this ever fail?

shouldnt it be is_cuda() ?

File: tests/data/test_idata.py (hunk 6 above)
Comment (fzimmermann89) [reply tests/data/test_idata.py:70]: () is missing

File: tests/data/test_kdata.py [hunk 8]
```diff
@@ -80,3 +80,47 @@ def test_KData_modify_header(ismrmrd_cart, field, value):
...
+    assert kcuda.traj.ky.is_cuda
+    assert kcuda.traj.kx.is_cuda
+
+
+@pytest.mark.cuda
+def test_KData_cuda(ismrmrd_cart):
+    """Move KData object to CUDA memory."""
+    k = KData.from_file(ismrmrd_cart.filename, DummyTrajectory())
+    kcuda = k.cuda()
+    assert kcuda.data.is_cuda
```
Comment (fzimmermann89) [inline tests/data/test_kdata.py:113]: see above...

File: src/mrpro/data/_KData.py (hunk 5 above)
Comment (ckolbPTB) [reply src/mrpro/data/_KData.py:258]: no .data is of type Tensor but .traj is of type Trajectory

File: src/mrpro/data/_KData.py (hunk 3 above)
Comment (ckolbPTB) [reply src/mrpro/data/_KData.py:215]: yes it is needed. mypy is also configured to throw an error if a type ignore is not needed

File: tests/data/test_idata.py (hunk 6 above)
Comment (ckolbPTB) [reply tests/data/test_idata.py:70]: is_cpu and is_cuda is a bool not a function

File: src/mrpro/data/_KData.py (hunk 3 above)
Comment (fzimmermann89) [reply src/mrpro/data/_KData.py:215]: which argment is the issue? 
I try to get this fixed in pytorch, but to looked ok to me.
` def to(self, device: Optional[DeviceLikeType] = None, dtype: Optional[_dtype] = None, non_blocking: _bool = False, copy: _bool = False, *, memory_format: Optional[torch.memory_format] = None) -> Tensor:`
is the typehint in pytorch. 
what am I missing?

File: src/mrpro/data/_KData.py (hunk 5 above)
Comment (fzimmermann89) [reply src/mrpro/data/_KData.py:258]: ah, thanks, got it

File: src/mrpro/data/_KData.py (hunk 3 above)
Comment (fzimmermann89) [reply src/mrpro/data/_KData.py:215]: ah, it got recently fixed in https://github.com/pytorch/pytorch/pull/113647

File: src/mrpro/operators/_FourierOp.py [hunk 9]
```diff
@@ -139,10 +139,10 @@ def __init__(
 
             self._omega = torch.stack(omega, dim=-2)
             self._fwd_nufft_op = KbNufft(
-                im_size=nufft_im_size, grid_size=grid_size, numpoints=numpoints, kbwidth=kbwidth
+                im_size=nufft_im_size, grid_size=grid_size, numpoints=numpoints, kbwidth=kbwidth, device=traj.kx.device
             )
             self._adj_nufft_op = KbNufftAdjoint(
-                im_size=nufft_im_size, grid_size=grid_size, numpoints=numpoints, kbwidth=kbwidth
+                im_size=nufft_im_size, grid_size=grid_size, numpoints=numpoints, kbwidth=kbwidth, device=traj.kx.device
```
Comment (fzimmermann89) [inline src/mrpro/operators/_FourierOp.py:143]: do we want to add a Ktraj.device ? Or more general, a Data.device?

**Comment by fzimmermann89** [comment]
lgtm.

three remarks: 

At some point in the near future for a new pytorch version some of the #ignore types will no longer be needed, but for older pytorch versions they will be.  this is something to keep in mind.

do we want to add a .device property to the data? instead of traj.kx.device it would allow traj.device

we should at some point set something up such that we don't have to write a _cuda test for all functions, but instead parameterize them to use either CPU (or even mixed devices)

**Review by fzimmermann89 — APPROVED** [review]

File: src/mrpro/operators/_FourierOp.py (hunk 9 above)
Comment (ckolbPTB) [reply src/mrpro/operators/_FourierOp.py:143]: A KData.device to me would only make sense if the entire object i.e. also the header is then on a certain device. For Ktraj it would make sense but I would postpone it because we still need something to verify that kx, ky, kz are actually on the same device, have some logic what happens if we assign mixed tensors to kx, ky, kz...
